lib/mayaUsd/ufe/UsdShaderAttributeHolder.cpp

```diff
@@ -0,0 +1,186 @@
+//
+// Copyright 2022 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "UsdShaderAttributeHolder.h"
+
+#include "UsdShaderAttributeDef.h"
+#include "Utils.h"
+
+#include <mayaUsd/utils/util.h>
+
+#ifdef UFE_V3_FEATURES_AVAILABLE
+#include <mayaUsd/base/tokens.h>
+#endif
+
+#include <pxr/base/tf/diagnostic.h>
+#include <pxr/base/tf/token.h>
+#include <pxr/base/vt/value.h>
+#include <pxr/pxr.h>
+
+namespace MAYAUSD_NS_DEF {
+namespace ufe {
+
+//------------------------------------------------------------------------------
+// UsdShaderAttributeHolder:
+//------------------------------------------------------------------------------
+
+UsdShaderAttributeHolder::UsdShaderAttributeHolder(
+    PXR_NS::UsdPrim                   usdPrim,
+    PXR_NS::SdrShaderPropertyConstPtr sdrProp,
+    PXR_NS::UsdShadeAttributeType     sdrType)
+    : UsdAttributeHolder(
+        usdPrim.GetAttribute(PXR_NS::UsdShadeUtils::GetFullName(sdrProp->GetName(), sdrType)))
+    , _usdPrim(usdPrim)
+    , _sdrProp(sdrProp)
+    , _sdrType(sdrType)
+{
+    // sdrProp must be valid at creation and will stay valid. usdPrim can be valid at creation and
+    // become invalid later.
+    PXR_NAMESPACE_USING_DIRECTIVE
+    TF_AXIOM(sdrProp && sdrType != PXR_NS::UsdShadeAttributeType::Invalid);
+}
+
+std::string UsdShaderAttributeHolder::isEditAllowedMsg() const
+{
+    if (_Base::isValid()) {
+        return _Base::isEditAllowedMsg();
+    } else if (_usdPrim) {
+        return std::string();
+    } else {
+        return "Editing is not allowed.";
+    }
+}
+
+std::string UsdShaderAttributeHolder::defaultValue() const
+{
+    if (_sdrProp->GetType() == PXR_NS::SdfValueTypeNames->Matrix3d.GetAsToken()) {
+        // There is no Matrix3d type in Sdr, so the MaterialX default value is not kept
+        return "0,0,0,0,0,0,0,0,0";
+    }
+#if PXR_VERSION < 2205
+    if (_sdrProp->GetType() == PXR_NS::SdfValueTypeNames->Bool.GetAsToken()) {
+        // Pre-22.05 there was no Boolean type in Sdr, so no default value
+        return "false";
+    }
+#endif
+    return UsdShaderAttributeDef(_sdrProp).defaultValue();
+}
+
+std::string UsdShaderAttributeHolder::nativeType() const { return _sdrProp->GetType(); }
```

Comment:
GetType() will properly return the MaterialX struct type (surfaceshader, BSDF, EDF, etc) so we will be able to filter connections correctly on generic attributes.